### ISSUE TITLE
Revisit .gitignore file treatment by balena build / balena push

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#1032]
[PR #950](https://github.com/balena-io/balena-cli/pull/950) implemented the feature described in [issue #889](https://github.com/balena-io/balena-cli/issues/889) but this has led to users reporting valid use cases where files should be uploaded for the build but not git-committed, hence listed in .gitnore files but not in .dockerignore files:

https://forums.balena.io/t/ignoring-dockerignore/2805/17

A user reported a situation where, for example, `.gitgnore` files present in subfolders like the Golang `vendor` folder (often consisting of github repo clones) cause the listed files to be excluded from the image build, when they shouldn't.

Consider alternative behaviours such as:

* Look at `.dockerignore` files only, not `.gitignore` (for both `balena build` and `balena push`). This would however be inconsistent with `git push` when used to trigger a build.
* Have `balena build` look at `.dockerignore` only, while `balena push` looks at both `.dockerignore` and `.gitignore`. (Inconsistency within the balena cli itself?)
* Having a command-line option to select which .ignore files to consider. By default, perhaps consider .dockerignore only.
* Having a `.balenaignore` file which could contain "import directives" to refer to the `.dockerignore` and `.gitignore` files that the user is interested in.
* Something else.

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_h503)

**Comment by magiconair** [comment]
My suggestion would be to leave the behavior as is since this is what most users are expecting but provide a flag to use only the one or the other or neither.

**Comment by matijse** [comment]
First of all, sorry for the long comment...

I ended up here because I had noticed unpredictable files being in/excluded with the `balena push` command. The documentation of this command doesn't mention what ignore files it looks at...

My setup:

In the main folder from where I run the `balena push` command, I have the following `docker-compose.yml` file. This file builds an application from a subfolder, which is a **git submodule**.

```
version: "2.1"

services:
  app:
    build:
      context: ./core-application/

  database:
    image: timescale/timescaledb:1.0.0-pg10
```
(relevant pieces shown)

I noticed the following cases, where I think the behaviour might be wrong:

Case 1: The `.gitignore` in the root of `core-application` contained the `vendor` folder like this: (with leading slash).

```
/vendor
```

This works (because `core-application` is its own git repository), the vendor map is not included in the git repository. However, in this setup, it is actually sent by the `balena push` command. Altough this is what I needed, I suspect this is wrong?

When I change the entry to:

```
vendor
```

The vendor folder is not copied with `balena push`.

Case 2:

Inside the `core-application` folder, I got a subfolder with logs. I need to keep the folder, but want to ignore the logs. Therefore I have the following structure:

```
/core-application/logs/some.log
/core-application/logs/.gitignore
/core-application/logs/keepme
```
logs/.gitignore file contents:
```
*
!keepme
```

Here, only the `keepme` file is added to git, but `balena push` ignores all files. 

If I instead remove the `.gitignore` file inside the `logs` directory, but in the root of the `core-application` repository, add the following to the `.gitignore`

```
logs/*.log
```
All logs are ignored, but the `keepme` file is kept, both in git and `balena push`.

--------------

As for my usage with this command, I would like to use this command in our deployment pipeline. Therefore I want to exclude the `vendor` folder from git, but not from sending to Balena.

Deployment would look like:
* Cloning repository with submodules
* In submodule(s) start some build process, for example `composer install`, generating `vendor` folder or other files ignored by git
* Use `balena push` to send all these assets to build server
* Use `.dockerignore` to exclude files from ending up in final image 

So in that case, both `.gitignore` and `.dockerignore` are not really suitable for looking for files to be ignored:

* `.gitignore` should contain `vendor` folder, as this should not be comitted to git
* `.dockerignore` should contain `composer.json` as this is not needed in the final image. It is however needed in the deployment pipeline to re-create the `vendor` folder.

Therefore, I would opt for a `.balenaignore` file where specific files might be excluded. I'm thinking this wil probably contain files that are present both in `.gitignore` and in `.dockerignore`...

**Comment by matijse** [comment]
Another interesting case:

My `.dockerignore` contains:

```
vendor/**/test*
vendor/**/tests*
vendor/**/Tests*
```

Which should ignore any test directory inside my dependencies. This works with a local `docker build`, but when I use `balena push`, it looks like any vendor pacakge that *has* tests, is ignored. Any package that has no tests, is included:

`vendor/symfony/var-dumper/Tests` is a folder inside the `var-dumper` package. 

A local `docker build` results in an image where `var-dumper` exists, and all subfolders are included, except for the `Tests` subfolder.

When I use `balena push`, I end up with an image where `var-dumper` is completely missing from the `vendor/symfony` folder. A few other packages do exist, for example `vendor/symfony/polyfill-ctype`, because that package has no tests folder.

**Comment by pdcastro** [comment]
> _First of all, sorry for the long comment..._

@matijse, first of all, thank you for the great comment! Illuminating.

**Comment by matijse** [comment]
I did some more digging, as I was still unable to use this command to succesfully deploy my application. (Based on Laravel 5.7). 

Laravel dependencies are packages inside the `vendor` directory. Each package is in its own directory and some packages also include a `.gitignore` file. (For example: https://github.com/symfony/var-dumper)

After some debugging, I noticed all packages I checked that didn't end up in the final image in Balena, had their own `.gitingore` file. When I locally removed the `.gitignore` file from its directory, the package got included. 

It turns out, that these gitignore files include the entry:

```
vendor/
```
which should only exclude any `vendor` folders below that directory. For some reason, the complete package directory is excluded, I think because this line "matches".

To be clear, not the whole `vendor` folder of my application is excluded, only the complete subfolder that contains a `.gitignore` which excludes `vendor/`.

```
core-application/vendor/symfony/polyfil-ctype  // Included
core-application/vendor/symfony/var-dumper     // Excluded
```

Are two packages I use. The first has no `.gitignore`, so that directory is completely included. The second has a `.gitignore` with `vendor/` entry. The complete `var-dumper` folder is excluded and does not exist in the final image.

When I remove the `.gitignore` or the `vendor/` line, the folder does get included.

**Comment by thgreasi** [comment]
Thanks for reporting this @matijse and for providing all those details.
I will to create a test case for this and fix the issue.

**Comment by matijse** [comment]
I was already looking into it a bit and created testcases that fail: https://gist.github.com/matijse/11604a42e750ad887fe7bab2084032b1

**Comment by thgreasi** [comment]
Oh that's awesome.
I would definitely love to review such a PR if you are willing to open one.

**Comment by matijse** [comment]
I only have the failing tests, do you want a PR for only that? I'm not sure I have time now to fix the issue...

**Comment by matijse** [comment]
I think I have a solution; the full relative path from the `basePath` was checked against every `.gitignore` file. For `.gitignore` files further down the tree, the relative path from that point needs to be used, in order not to accidentally match parts in the parent path.

**Comment by magiconair** [comment]
FWIW, I have a Go project with a `vendor` folder which has a `vendor/*/.gitignore` file which excludes `/vendor/` which makes sense for the library itself.

`balena build` will exclude the entire `vendor/` folder which fails the build. The only way around this is to delete all `.gitignore` files before the build and restore them later. This works with Docker.

I've created a sample project to demo this:

https://github.com/magiconair/balena-gitignore

**Comment by matijse** [comment]
Thanks for the sample @magiconair ! I have exactly the same, and for now just delete all .gitignore files. (`find . -type f -name .gitignore -exec rm {} \;`) 

I dug a bit deeper in the code, and to fix it properly, requires some work. It uses the [ignore](https://github.com/kaelzhang/node-ignore) npm package, which should handle only one `.gitignore` file at a time, and all files checked should be relative to that `.gitignore` file. 

The current balena-cli checks for each file if it is ignored, but adds rules from all relevant `.gitignore` files to the rule-set ([source](https://github.com/balena-io/balena-cli/blob/master/lib/utils/ignore.ts#L114-L120)). This needs to be changed, so for each `.gitignore` file a ignore handler is created, and files are filtered when one of the `.gitingore` rule-sets exclude them.

**Comment by magiconair** [comment]
@matijse quick tip: `find . -type f -name .gitignore -delete` does the same thing.

**Comment by pdcastro** [comment]
A different issue arising from the current treatment of the `.gitignore` file was reported in /issues/1092. A `.balena` line was added to the `.gitignore` file, and it resulted in a cryptic `".balena/qemu-execve: no such file or directory"` error when running `balena build`.  As I understand it, the CLI creates a `.balena` folder in an in-memory tar stream that it sends to the docker daemon. It appears that the CLI is allowing entries in the `.gitignore` file to hide files and folders that the CLI itself creates in memory. This sounds like a bug that should be addressed here as well.

**Comment by mz8i** [comment]
I just ran into a problem caused by this and wanted to pitch in with my view.
Our situation is that we have a node/typescript project. The Dockerfile itself does not run the transpilation, it just relies on copying the JS source files from a build folder which contains the results of transpilation.

The problem I faced occurred when I was running `balena build` (with the `--emulated` flag - not sure if it actually matters). The docker builder was complaining about the build folder supplied as argument to COPY not being present. After trawling through the resin-multibuild source I discovered it's taking .gitignore into consideration. The build folder is obviously in .gitignore in our case. Using `balena deploy` worked fine, with the build running with no problems on a local Raspberry Pi.

I personally don't have an opinion about `balena push`, where it's probably reasonable to take use rules from .gitignore - but for `balena build`, which is essentially a local docker image build with added emulation and other useful mechanisms, any source control mechanism has no relevance whatsoever. If I just try to run a docker image build locally, the only thing that matters is that the files referred to in the `Dockerfile` are present locally. Whether they are being tracked by version control or not is not important. Taking .dockerignore into consideration when copying files into the temp directory probably makes sense, though.

To sum up, I suggest .gitignore should not be used in the local `balena build`

**Comment by wlisac** [comment]
I also hit this issue today. A flag to customize the behavior would be helpful 🙂

**Comment by r4space** [comment]
Also just discovered this issue...

FWIW I'd second @matijse proposal for a .balenaignore (along with .gitignore and .dockerignore being respected by their own tools)  Not really that much more work for user (do nothing other than cp .gitignore if not needed) and would give full fine grained control

One simple use case I'd use .balenaignore for would be easy segregation of documentation folders that I want in the github repo and tracked but obviously have no reason to deploy them to devices.  I know there are other git based mechanisms for achieving this but this would be a nice and simple feature along with  ensuring clarity and solving all the unanticipated issues we've all had here 

Either way - some more obvious documentation/early warning on what balena cml looks at should be created sooner than later

**Comment by zrzka** [comment]
I just hit this issue as well. My project does contain `nvidia` folder, which contains JetPack SDK. I need this folder (JetPack SDK files) to build my project, but I can't provide them for others due to licensing. Hence the `/nvidia` folder in the `.gitignore` file.

I personally second the `.balenaignore` proposal. We can't never make everyone happy with `.gitignore` & `.dockerignore`, ... Let these files be respected by their own tools and introduce a new one.

**Comment by eczajk1** [comment]
Note that this issue now impacts local development as well with the changes to [Local Mode](https://www.balena.io/blog/living-on-the-edge-with-livepush-and-other-cli-improvements/#localmode).

**Comment by zrzka** [comment]
Hmm, I slept on this and I'm not sure that the `.balenaignore` is the right way. What use cases do we have here:

| Git versioned | Required for build | Where should it go |
| -- | -- | -- |
| ✓ | ✓ | ✗ |
| ✓ | ✗ | `.dockerignore` |
| ✗ | ✓ | `.gitignore` [1] |
| ✗ | ✗ | ✗ |

[1] The only problem here is that the `git push balena master` won't work and user has to use `balena push`. And I think it's fine, because it's a very specific case and it will never work with `git push balena master` as the file isn't versioned.

Let me correct my proposal:

* `balena push` should ignore `.gitignore`
* `balena push` should look into the `.dockerignore` files only

This proposal is also future compatible in a way that if the `git push balena master` will be banned (only `balena push` will work), there will be no need to change anything.

From this point of view, I think that the `.balenaignore` will only shadow `.dockerignore` behavior. Am I wrong and am I overlooking something?

**Comment by pdcastro** [comment]
@zrzka, thanks for this analysis. An issue with the proposal of "ignoring `.gitignore`" is that `balena push` is usually "promoted" as a _"drop-in replacement for git push",_ as this removes obstacles for users to switch from `git push` to `balena push`. Ideally we would have a situation where _"if it works with git push, then it works with balena push too"_ (so `balena push` shouldn't upload files that `git push` excludes), even if the reverse isn't necessarily true (i.e., _"git push offers a subset of the features of balena push")._

**Comment by zrzka** [comment]
@pdcastro I see, but my situation is that it doesn't work with git push (and never will because of .gitignore).

I'd like to point out that balena push is like git push & docker build together, not just git push. In other words, docker build doesn't take care of .gitignore files and balena push shouldn't either. But it's my personal opinion and I understand why it is done / promoted in this way.

**Comment by zrzka** [comment]
Maybe the right way here is to use [exclude](https://help.github.com/en/articles/ignoring-files#explicit-repository-excludes)? It's local only, git related, balena push doesn't read / use it, enough for development. And as others are going to clone & build only ... Thinking aloud here ...

**Comment by zrzka** [comment]
If balena push should stay "git compatible", what about _balena push ... --ignored_? Push even git ignored files? Same as _git ls-files --ignored_, which lists even ignored files. Same flag, same behavior.

**Comment by chances** [comment]
I hit this issue today while I'm cross compiling a Go project before `balena push` and want my compiled binary, which is ignored by `.gitignore`, copied to the Docker image. Compiling the Go project on my Pi device directly is _far_ too slow.

This is completely blocking my workflow. How can I work around this limitation?

**Comment by pdcastro** [comment]
> _I want my compiled binary, which is ignored by .gitignore, copied to the Docker image. [...] This is completely blocking my workflow. How can I work around this limitation?_

In terms of "unblocking and getting around", perhaps:
```
cp .gitignore .gitignore.original
cp .gitignore .gitignore.push
```
Then edit `.gitignore.push` to remove the compiled binary name, and create this simple `push.sh` script:
```
cp .gitignore.push .gitignore
balena push ...
cp .gitignore.original .gitignore
```

**Comment by pdcastro** [comment]
An earlier comment in this thread described an example where a file needed to be uploaded by the CLI (balena push) to the balenaCloud builder, but the file should not make its way to the final built image. Another earlier comment pointed out different needs between the `balena build` and `balena push` commands. Thinking about it, there may be a case for having the balena CLI and balenaCloud builder look a `.dockerignore` and `.balenaignore` file depending on the circumstances. Here's a suggestion, starting with a recap of each command's behaviour:

| Command | Behaviour |
| -- | -- |
| git push<br>balena push appName | Upload files to the balenaCloud builder. |
| balena push IPaddress<br>balena build<br>balena deploy -b | Upload files to a local or remote docker daemon. |

Proposal:

| File | Command | File Purpose |
| -- | -- | -- |
| `.gitignore` | git push | Files listed in `.gitignore` are not uploaded to the balenaCloud build server. |
| `.gitignore` | balena push<br>balena build<br>balena deploy -b | Not used. Users migrating from `git push` to `balena push` should duplicate (copy) `.gitignore` as `.balenaignore`.  |
| `.balenaignore` | git push | Not used |
| `.balenaignore` | balena push appName | Exclude files from being uploaded to the balenaCloud builder. The builder then looks at a `.dockerignore` file (if any) to further exclude files from the final image. |
| `.balenaignore` or `.dockerignore` | balena push IPaddress<br>balena build<br>balena deploy -b | Exclude files from being sent to the local or remote docker daemon -- like the `docker build` command uses `.dockerignore`. If both `.balenaignore` and `.dockerignore` are present, the CLI uses `.balenaignore` only. |
| `.dockerignore` | git push<br>balena push appName | Not used on the client side, but the `.dockerignore` file itself is uploaded to the balenaCloud builder. The builder uses it to decide what files (that have been uploaded) should be excluded from the final built image. |

**Comment by zrzka** [comment]
Here's my question and concern for the `.balenaignore` & `balena push appName` part:

> Exclude files from being uploaded to the balenaCloud builder. The builder then looks at a `.dockerignore` file (if any) to further exclude files from the final image.

Is there any use case for this? Upload a file to the builder (not listed in `.balenaignore`), which is going to ignore it (listed in `.dockerignore`)?

If there isn't any use case for this, can we remove the complexity with three `.*ignore` files to just two by removing `.balenaignore` and using just `.dockerignore`? See below.

Sorry for nitpicking, I'm just trying to understand the purpose of this `.balenaignore` file.

---

| File | Command | File Purpose |
| -- | -- | -- |
| `.gitignore` | git push | Files listed in `.gitignore` are not uploaded to the balenaCloud build server. |
| `.gitignore` | balena push<br>balena build<br>balena deploy -b | Not used. Users migrating from `git push` to `balena push` should duplicate (copy) `.gitignore` as `.dockerignore`.  |
| `.dockerignore` | git push | Not used |
| `.dockerignore ` | balena push appName | Exclude files from being uploaded to the balenaCloud builder. The builder then looks at a `.dockerignore` file (if any) to further exclude files from the final image. |
| `.dockerignore` | balena push IPaddress<br>balena build<br>balena deploy -b | Exclude files from being sent to the local or remote docker daemon -- like the `docker build` command uses `.dockerignore`. |
| `.dockerignore` | git push<br>balena push appName | Not used on the client side, but the `.dockerignore` file itself is uploaded to the balenaCloud builder. The builder uses it to decide what files (that have been uploaded) should be excluded from the final built image. |

**Comment by mz8i** [comment]
I was also wondering after reading this, how is `.balenaignore` different, specifically, from `.dockerignore`, and is it needed at all?

**Comment by wlisac** [comment]
I've been thinking about this quite a bit lately and I'm really hoping we can come to a solution.

I might be missing something, but I also don't understand the use case for `.balenaignore`.

I think one of @zrzka's earlier posts summarize what I'm looking for:

> * `balena push` should ignore `.gitignore`
> * `balena push` should look into the `.dockerignore` files only

In other words, something like this:

| File             | Command                                         | File Purpose                                                                                                     |
|------------------|-------------------------------------------------|------------------------------------------------------------------------------------------------------------------|
| `.gitignore`     | git push                                        | Files listed in `.gitignore` are not uploaded to the balenaCloud build server.                                   |
| `.dockerignore`  | git push                                        | Not used on the client side. The builder uses it to decide what files should be excluded from the build context. |
| `.gitignore`     | balena push<br>balena build<br>balena deploy -b | Not used. Users migrating from `git push` to `balena push` should duplicate `.gitignore` to `.dockerignore`.     |
| `.dockerignore ` | balena push<br>balena build<br>balena deploy -b | Excludes files from being sent to the balenaCloud builder or docker daemon.                                      |


Respecting `.dockerignore` when using `git push` also lines up with what's mentioned in https://github.com/balena-io/docs/issues/963.

**Comment by wlisac** [comment]
@pdcastro any thoughts on that above proposal?

**Comment by pdcastro** [comment]
> pdcastro any thoughts on that above proposal?

Yes, thanks for asking! We have recently discussed it internally and made a decision, and I am currently implementing it. Initially we will change the CLI only, not touching the balenaCloud builders. The result in then quite simple:

* `git push`: unchanged. It looks at the `.gitignore` only.
* All `balena` commands (build, deploy, push): they will look at the `.dockerignore` file only, to filter out what to upload to the build servers.
* No `.balenaignore` file.

Because it is a breaking change, and also to allow for further testing and feedback, the plan is to initially introduce a command-line option that enables the new behaviour. If the command-line option is not used, the old behaviour applies but a deprecation warning message is printed. In a future major release, the new behaviour then becomes the only behaviour, and the flag becomes redundant.

**Comment by wlisac** [comment]
@pdcastro Nice. This sounds great! Looking forward to trying it out.

**Comment by wlisac** [comment]
@pdcastro Is there any update you can share on this?

**Comment by pdcastro** [comment]
@wlisac, indeed I came across an unexpected issue: the [zeit/dockerignore](https://www.npmjs.com/package/@zeit/dockerignore) library (3rd party) has bugs affecting Windows paths. I've raised [this issue](https://github.com/zeit/dockerignore/issues/15), and some later tests suggest that the problem may be even worse, affecting any patterns containing a path separator (whether forward slash or backslash) on Windows. [Their Windows build is failing](https://ci.appveyor.com/project/zeit/dockerignore).

The problem now is that, if we move away from .gitignore towards .dockerignore as planned here, despite the bugs in the zeit/dockerignore library, this would be break the builds of some of our Windows users who currently rely on .gitignore. So we need to fix those bugs, or find another library. I was planning to try and fix the bugs in the zeit/dockerignore library, but then was required to shift my attention to other issues.

So a **call for help:** if anyone would be able to fix the zeit/dockerignore library on Windows (and ideally get their AppVeyor Windows build tests to pass), it would unblock progress for this issue.

**Comment by wlisac** [comment]
@pdcastro Thanks for the update :)

**Comment by bekriebel** [comment]
I'm evaluating Balena right now and just hit this issue with a go project. If [this project](https://github.com/sirupsen/logrus) is included in a `vendor` directory, the way the .gitignore is being read is making the whole logrus folder not get included and failing the build. I'm a bit worried to see that this has been an issue for a year and a half and it hasn't been resolved yet.

**Comment by pdcastro** [comment]
@bekriebel, sorry for the delay in addressing this issue. To some extent, it is blocked by another issue with a dependency package (https://github.com/balena-io/balena-cli/issues/1032#issuecomment-528357887), though of course any of us can work on that other issue and I have been meaning to do it for a while. We have a public ordered backlog for balena-cli issues, and this issue is very close to the top (currently 3rd): 
https://github.com/balena-io/balena-cli/milestone/2

So hopefully it won't be that much longer until progress is made. The balena team will pick it up for sure, but also I could add that it is a large open source project and community contributions are very welcome. :-)

**Comment by pdcastro** [comment]
A glimmer of light at the end of the tunnel: good progress has been made in addressing the issues with the `zeit/dockerignore` dependency that were blocking the work on this issue proper: https://github.com/zeit/dockerignore/pull/16
There is still more work to do, but it is progressing.

**Comment by pavlospt** [comment]
@pdcastro do you need any help regarding the AppVeyor CI pipeline issues?

**Comment by pdcastro** [comment]
@pavlospt, thanks for the offer, it looks like it was just a matter of adding the `.appveyor.yml` file to that same PR: https://ci.appveyor.com/project/zeit/dockerignore/builds/32073705

**Comment by pavlospt** [comment]
Yeap that was my first observation. That there was no appveyor.yml :) happy to see it solved!

**Comment by pdcastro** [comment]
It's here! CLI release v11.34.0 introduces a new flag for the push / build / deploy commands: `--nogitignore` or `-G`. Quoting `balena help push`:
```
$ balena help push
...
DOCKERIGNORE AND GITIGNORE FILES
By default, both '.dockerignore' and '.gitignore' files are taken into account
in order to prevent files from being sent to the balenaCloud builder or Docker
or balenaEngine (balenaOS device).

However, this behavior has been DEPRECATED and will change in an upcoming major
version release. The --nogitignore (-G) option should be used to enable the new
behavior already now. This option will cause the CLI to:

* Disregard all '.gitignore' files at the source directory and subdirectories,
  and consider only the '.dockerignore' file (if any) at the source directory.
* Consequently, allow files to be sent to balenaCloud / Docker / balenaEngine
  even if they are listed in '.gitignore' files (a longstanding feature request).
* Use a new '.dockerignore' parser and filter library that improves compatibility
  with "docker build" and fixes several issues (mainly on Windows).
* Prevent a warning message from being printed.

When --nogitignore (-G) is provided, a few "hardcoded" dockerignore patterns are
also used and "merged" (in memory) with the patterns found in the '.dockerignore'
file (if any), in the following order:

    **/.git
    < user's patterns from the '.dockerignore' file, if any >
    !**/.balena
    !**/.resin
    !**/Dockerfile
    !**/Dockerfile.*
    !**/docker-compose.yml

If necessary, the effect of the '**/.git' pattern may be modified by adding
"counter patterns" to the '.dockerignore' file, for example '!service1/.git'.
For documentation on pattern format, see:
- https://docs.docker.com/engine/reference/builder/#dockerignore-file
- https://www.npmjs.com/package/@balena/dockerignore
```
And similar for `balena help build` and `balena help deploy`.
Give it a try: feedback is still welcome! The plan is for the `--nogitignore` option to become the default behavior in the next major version release (v12).

Some notes:
* The matching pattern format is not the same between `.gitignore` and `.dockerignore` files. A brief summary of the differences can be found at: https://www.npmjs.com/package/@balena/dockerignore#whats-different-from-ignore-
* While multiple `.gitignore` files are allowed (in subdirectories), this is not the case for `.dockerignore`: only one `.dockerignore` file is allowed, at the "project root".
  * This is the case even for multi-service / multi-container applications: the one `.dockerignore` file should be at same level as the `docker-compose.yml` file (if any), and service names may be listed in the one `.dockerignore` file. The `**` pattern may also be used to recursively match multiple subdirectories, and may be prefixed by the service name for this to effectively apply per service.
* balena application deployments with `git push` (instead of `balena push`) is unchanged: `git push` disregards `.dockerignore` files.
* The balenaCloud builder is also unchanged. Like the Docker daemon or balenaEngine, the balenaCloud builders disregard `.dockerignore` or `.gitignore` files. All filtering happens at the client side: git client, balena CLI, or docker CLI.

**Comment by pdcastro** [comment]
Closing this issue with [the release of CLI v12](https://github.com/balena-io/balena-cli/wiki/CLI-v12-Release-Notes), where the `--nogitignore` option became the default behavior. CLI v12.1.0 also introduced a `--gitignore` (`-g`) option for v11 compatibility, in case your project isn't ready. The plan is to remove this option in CLI v13.

But it looks like it is not the end of the story just yet: see #1870 [Consider a .dockerignore file per service for docker-compose compatibility](https://github.com/balena-io/balena-cli/issues/1870)

**Comment by pdcastro** [comment]
CLI v12.2.0 introduces the `--multi-dockeringore` (`-m`) option that allows each service of a microservices / multicontainer app to define its own `.dockerignore` file, in the way that the "standard" `docker-compose` tool works:
```
$ balena help push
...
The --multi-dockerignore (-m) option may be used with microservices (multicontainer)
applications that define a docker-compose.yml file. When this option is used,
each service subdirectory (defined by the `build` or `build.context` service
properties in the docker-compose.yml file) is filtered separately according to
a .dockerignore file defined in the service subdirectory. If no .dockerignore
file exists in a service subdirectory, then only the default .dockerignore
patterns (see below) apply for that service subdirectory.

When the --multi-dockerignore (-m) option is used, the .dockerignore file (if
any) defined at the overall project root will be used to filter files and
subdirectories other than service subdirectories. It will not have any effect
on service subdirectories, whether or not a service subdirectory defines its
own .dockerignore file. Multiple .dockerignore files are not merged or added
together, and cannot override or extend other files. This behavior maximises
compatibility with the standard docker-compose tool, while still allowing a
root .dockerignore file (at the overall project root) to filter files and
folders that are outside service subdirectories.
...
```
The `--multi-dockerignore` (`-m`) and the `--gitignore` (`-g`) options are mutually exclusive. The next major CLI release (v13) will drop the `--gitignore` (`-g`) option and, depending on feedback, will make the `--multi-dockerignore` (`-m`) option the default behavior, so that the option does not need to be specified.
